warehouse/macaroons/auth_policy.py

```diff
@@ -0,0 +1,120 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pyramid.authentication import CallbackAuthenticationPolicy
+from pyramid.interfaces import IAuthenticationPolicy, IAuthorizationPolicy
+from pyramid.security import Denied
+from pyramid.threadlocal import get_current_request
+from zope.interface import implementer
+
+from warehouse.cache.http import add_vary_callback
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.macaroons.services import InvalidMacaroon
+
+
+def extract_http_macaroon(request):
+    """
+    A helper function for the extraction of HTTP Macaroon from a given request.
+    Returns either a ``None`` if no macaroon could be found, or the byte string
+    that represents our Macaroon.
+    """
+    authorization = request.headers.get("Authorization")
+    if not authorization:
+        return None
+
+    try:
+        auth_method, auth = authorization.split(" ", 1)
+    except ValueError:
+        return None
+
+    if auth_method.lower() != "macaroon":
```

Comment:
Just noting that this will have to change if we want to make it capable of being used with existing clients without modifications to pull out the password from the client. Not sure what we should expect for username in that case, we could either expect the username to match the macaroon or ignore the username? Not sure if there's a better option.

Another option is to support *both* this header and the basic auth shim, and try to guide people over to using this header instead. Not sure if that is worthwhile or not.

Reply:
Hmm. I'm not sure about worthwhileness either, but my intuition is that this should *eventually* become its own header entirely, rather than piggybacking on `Authentication: basic` (which it'll have to do for this rollout).

Here's one idea: we could support both `Authorization: macaroon <macaroon>` and `Authorization: basic <token>`, where `<token>` is `base64("macaroon:" . <macaroon>)`. That results in a pretty chunky header since we end up encoding the macaroon twice, but it preserves compatibility and gives us a "username" that clients could enter without making any changes. Thoughts?

Reply:
Ended up going with the nested `basic` idea, with the "username" as `@token` to distinguish it from a real username.